### ISSUE TITLE
Issue with torch.utils.data.DataLoader 'shuffle=True' 

### ISSUE THREAD

**Issue by dbl001** [pytorch/pytorch#83270]
### 🐛 Describe the bug

When 'shuffle=True' the accuracy of my Resnet50 predictions drops dramatically.
The problem occurs with 'mps' backend as well as the 'cpu' device. With 'batch_size = 1' there is no problem.
When the 'batch_size' is > 1, it appears the order of the predictions (-vs- labels) is sometime reversed.
I am using Resnet50 with the 'shared_interest' Jupyter notebook on github.

https://github.com/mitvis/shared-interest/blob/main/shared_interest/examples/shared_interest_example.ipynb

Here is my test code:
```
import json
import matplotlib.pyplot as plt
import numpy as np
import os
import re
import PIL.Image
from sklearn.metrics import accuracy_score
import torch
import torchvision.models as models
from torchvision.models import resnet50, ResNet50_Weights
import torchvision.transforms as transforms
from tqdm import tqdm
import pandas as pd
from torch.utils.data import RandomSampler
    
from shared_interest.datasets.imagenet import ImageNet
from shared_interest.shared_interest import shared_interest
from shared_interest.util import flatten, binarize_std
from interpretability_methods.vanilla_gradients import VanillaGradients

def load_model_from_pytorch(architecture, pretrained):
    model = models.__dict__[architecture](weights=ResNet50_Weights.IMAGENET1K_V1)
    #model = models.__dict__[architecture](pretrained=pretrained)
    return model

model = load_model_from_pytorch('resnet50', pretrained=True)
#device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
device = 'cpu'
model = model.eval().to(device)

imagenet_dir = '/Users/davidlaxer/ImageNet/ILSVRC'
image_dir = os.path.join(imagenet_dir, 'Data/CLS-LOC/train')
annotation_dir = os.path.join(imagenet_dir, 'Annotations/CLS-LOC/train')

df = pd.read_fwf('/Users/davidlaxer/ImageNet/LOC_synset_mapping.txt', sep=' ', index_col=False, header=None, names=['WordNetId', 'Class'], set_index='WordNetId')
df['row_num'] = np.arange(len(df))

# ImageNet transforms.
image_transform = transforms.Compose([transforms.Resize(256),
                                      transforms.CenterCrop(224),
                                      transforms.ToTensor(),
                                      transforms.Normalize(mean=[0.485, 0.456, 0.406], 
                                                           std=[0.229, 0.224, 0.225]),
                                     ])

ground_truth_transform = transforms.Compose([transforms.ToPILImage(),
                                             transforms.Resize(256, PIL.Image.NEAREST),
                                             transforms.CenterCrop(224),
                                             transforms.ToTensor()])

reverse_image_transform = transforms.Compose([transforms.Normalize(mean=[0, 0, 0], 
                                                                   std=[4.3668, 4.4643, 4.4444]),
                                              transforms.Normalize(mean=[-0.485, -0.456, -0.406], 
                                                                   std=[1, 1, 1]),
                                              transforms.ToPILImage(),])

dataset = ImageNet(image_dir, annotation_dir, image_transform, ground_truth_transform)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=2, shuffle=True, num_workers=0, pin_memory=True)

for image, ground_truth, label in dataset:
    if ground_truth == None:
        continue
    fig, ax = plt.subplots(nrows=1, ncols=2)
    ax[0].imshow(reverse_image_transform(image))
    ax[1].imshow(ground_truth)
    type(image)
    break

saliency_method = VanillaGradients(model)

def run(model, dataloader, saliency_method, stop_after=None):
    """
    Runs the model through the data in the dataloader and computes the 
    predictions, saliency, and Shared Interest scores.
    
    Args:
    model: pytorch model to evaluate.
    dataloader: dataloader to evaluate the model on. Should output an image,
        ground truth, and label for each index.
    saliency_method: the InterpretabilityMethod to use to compute Shared
        Interest.
    stop_after: optional integer that determins when to stop the evaluation.
        Used here for efficiency in the notebook. If None, entire process will
        run.
    """
    accuracy = 0
    total_shared_interest_scores = {'iou_coverage': np.array([]),
                                    'ground_truth_coverage': np.array([]),
                                    'saliency_coverage': np.array([]),}
    total_predictions = np.array([])
    total_saliency_masks = np.array([[[]]])
    for i, (images, ground_truth, labels) in enumerate(tqdm(dataloader)):
        with torch.no_grad():
            images = images.to(device, torch.float32)
            ground_truth = ground_truth.numpy()
            print(labels)
            df1 = pd.DataFrame(labels, columns=['WordNetId'])
            df2 = df.merge(df1)['row_num']
            labels = df2.to_numpy()
        
            # Compute model predictions
            output = model(images)
            predictions = torch.argmax(output, dim=1).detach().cpu().numpy()
            total_predictions = np.concatenate((total_predictions, predictions))
            
            # Update metrics
            print(accuracy_score(labels, predictions))
            print("Labels:\t\t", labels)
            #print("Output:\t\t", output)
            print("Predictions:\t", predictions, "\n")
            accuracy += accuracy_score(labels, predictions)
        
        # Compute saliency
        saliency = flatten(saliency_method.get_saliency(images))
        saliency_masks = binarize_std(saliency)
        if i == 0: 
            total_saliency_masks = saliency_masks
        else:
            total_saliency_masks = np.concatenate((total_saliency_masks, saliency_masks))
        
        # Compute Shared Interest scores
        for score in total_shared_interest_scores:
            shared_interest_scores = shared_interest(ground_truth, saliency_masks, score=score)
            total_shared_interest_scores[score] = np.concatenate((total_shared_interest_scores[score], shared_interest_scores))

        # Stop early for this example notebook
        if stop_after and stop_after == i:
            break
    accuracy /= i + 1
    print('Accuracy: %.2f' %accuracy)
    return total_saliency_masks, total_shared_interest_scores, total_predictions

saliency_masks, shared_interest_scores, predictions = run(model, dataloader, saliency_method, stop_after=10)

```
Here is example output using device='cpu', 'batch_size=2', 'shuffle=True':
```
/Users/davidlaxer/anaconda3/envs/AI-Feynman/bin/python /Users/davidlaxer/shared-interest/test3.py 
/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/torchvision-0.14.0a0+e75a333-py3.9-macosx-10.9-x86_64.egg/torchvision/transforms/transforms.py:329: UserWarning: Argument 'interpolation' of type int is deprecated since 0.13 and will be removed in 0.15. Please use InterpolationMode enum.
  warnings.warn(
  0%|          | 0/272274 [00:00<?, ?it/s]('n04371774', 'n01667778')
0.0
Labels:		 [ 36 843]
Predictions:	 [843  36] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 1/272274 [00:00<54:46:51,  1.38it/s]('n02823428', 'n01688243')
0.0
Labels:		 [ 43 440]
Predictions:	 [440  43] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 2/272274 [00:01<54:54:49,  1.38it/s]('n03447447', 'n02782093')
0.0
Labels:		 [417 576]
Predictions:	 [576 417] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 3/272274 [00:02<56:20:51,  1.34it/s]('n02795169', 'n04517823')
1.0
Labels:		 [427 882]
Predictions:	 [427 882] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 4/272274 [00:02<55:20:43,  1.37it/s]('n01797886', 'n03476991')
0.5
Labels:		 [ 82 585]
Predictions:	 [ 82 903] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 5/272274 [00:03<55:40:11,  1.36it/s]('n02268443', 'n02119789')
0.0
Labels:		 [278 319]
Predictions:	 [319 278] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 6/272274 [00:04<56:07:05,  1.35it/s]('n03450230', 'n03825788')
1.0
Labels:		 [578 680]
Predictions:	 [578 680] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 7/272274 [00:05<57:19:28,  1.32it/s]('n03710637', 'n02493509')
0.0
Labels:		 [380 638]
Predictions:	 [638 380] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 8/272274 [00:05<56:54:56,  1.33it/s]('n03207743', 'n01514859')
0.0
Labels:		 [  8 533]
Predictions:	 [533   8] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 9/272274 [00:06<59:30:00,  1.27it/s]('n02701002', 'n04465501')
1.0
Labels:		 [407 866]
Predictions:	 [407 866] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
  0%|          | 10/272274 [00:07<66:49:00,  1.13it/s]('n03032252', 'n02815834')
0.0
Labels:		 [438 498]
Predictions:	 [498 438] 

/Users/davidlaxer/anaconda3/envs/AI-Feynman/lib/python3.9/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
Accuracy: 0.32
  0%|          | 10/272274 [00:08<68:00:10,  1.11it/s]

Process finished with exit code 0

```

### Versions

```
 % python collect_env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (x86_64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.3
Libc version: N/A

Python version: 3.9.12 (main, Jun  1 2022, 06:36:29)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] pytorch-lightning==1.7.0
[pip3] torch==1.13.0a0+gitf4ee374
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.14.0a0+e75a333
[conda] blas                      1.0                         mkl    anaconda
[conda] mkl                       2021.4.0           hecd8cb5_637    anaconda
[conda] mkl-include               2022.0.0           hecd8cb5_105    anaconda
[conda] mkl-service               2.4.0            py39h9ed2024_0    anaconda
[conda] mkl_fft                   1.3.1            py39h4ab4a9b_0    anaconda
[conda] mkl_random                1.2.2            py39hb2f4e1b_0    anaconda
[conda] numpy                     1.22.3           py39h2e5f0a9_0    anaconda
[conda] numpy-base                1.22.3           py39h3b1a694_0    anaconda
[conda] pytorch                   1.12.1                  py3.9_0    pytorch
[conda] pytorch-lightning         1.7.0                    pypi_0    pypi
[conda] torch                     1.13.0a0+git7d17e3b          pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.14.0a0+e75a333          pypi_0    pypi
```

cc @SsnL @VitalyFedyunin @ejguan @NivekT

**Comment by ejguan** [comment]
I don't think this Error comes from DataLoader. It's apparently not all of labels are reversed in your log. I believe the Error comes from your custom pre-processing logic. I am closing this Issue for now. You might want to open an issue from the shared_interest repo not in PyTorch Core.

And, skimming through your code, I guess the culprit is from here:
```py
df2 = df.merge(df1)['row_num']
labels = df2.to_numpy()
```
When you merge two DataFrames, you are using the order from `df`. Since you have shuffle enabled, you might want to do merge from right hand side.
